internal/webhook/external/github.go

```diff
@@ -0,0 +1,152 @@
+	// environment variable is empty.
+	errSecretUnset = errors.New("secret is unset")
+)
+
+// githubHandler handles push events for github.
+// After the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func githubHandler(c client.Client) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		secret, ok := os.LookupEnv("GH_WEBHOOK_SECRET")
+		if !ok {
+			logger.Error(errSecretUnset, "environment misconfiguration")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					// keep it vague, no need to leak config details
+					errors.New("unexpected server error"),
+					http.StatusInternalServerError,
+				),
+			)
+			return
+		}
+
+		signature := r.Header.Get(gh.SHA256SignatureHeader)
+		if signature == "" {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					errMissingSignature,
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
+
+		bodyBytes, err := xhttp.LimitRead(r.Body)
+		if err != nil {
+			xhttp.WriteErrorJSON(w, err)
+			return
+		}
+
+		if err = gh.ValidateSignature(
+			signature,
+			bodyBytes,
+			[]byte(secret),
+		); err != nil {
+			logger.Error(err, "failed to validate signature")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					// same here
+					errors.New("unauthorized"),
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
+
+		eventType := r.Header.Get("X-GitHub-Event")
+		if eventType != "push" {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("only push events are supported"),
+					http.StatusNotImplemented,
+				),
+			)
+			return
+		}
+
+		e, err := gh.ParseWebHook(eventType, bodyBytes)
+		if err != nil {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to parse webhook event: %w", err),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		pe, ok := e.(*gh.PushEvent)
+		if !ok {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("only push events are supported"),
+					http.StatusNotImplemented,
+				),
+			)
+			return
+		}
+
+		repo := *pe.Repo.HTMLURL
+		logger.Debug("source repository retrieved", "name", repo)
+		result, err := refresh(ctx, c, logger, repo)
+		if err != nil {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(err, http.StatusInternalServerError),
+			)
+			return
+		}
+
+		logger.Debug("execution complete",
+			"total", result.totalWarehouses,
+			"num-failures", result.numFailures,
+		)
+
+		if result.numFailures > 0 {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to refresh %d of %d warehouses",
+						result.numFailures,
+						result.totalWarehouses,
+					),
+					http.StatusInternalServerError,
+				),
+			)
```

Comment:
Because we do not typically want to expose too much information about internal server errors, the `WriteErrorJSON()` function is suppressing the message on 500s.

Since we do want this particular message to be relayed back to the client, we should use `WriteResponseJSON()` instead since it gives us more control:

```suggestion
			xhttp.WriteResponseJSON(
				w,
				http.StatusInternalServerError,
				struct{
					Error string `json:"error"`
				}{
					Error: fmt.Errorf(
						"failed to refresh %d of %d warehouses",
						result.numFailures, result.totalWarehouses,
					),
				}
			)
```